### PR TITLE
Suggestion to improve tracing in nats-server through wrapping errors

### PR DESCRIPTION
**Pull request by matthiashanel**
This is primarily a suggestion to enhance tracing. I'll incorporate any feedback or discard it if it does not add as much value as I thing it could.

I noticed a slack discussion involving ErrMaxControlLine. Looking up the code I noticed that the trace for ErrMaxControlLine could/should be improved.
Instead of addressing this one particular situation by returning a different error with more context I sought to find a more general solution.

A general solution that:
1: allows the continued use of error constants
2: can be added gradually
3: only shows the added context where needed (in tracing, not when sending errors to clients)

Go 1.13 adds support for wrapping errors.
Instead of wrapping errors in more specific types I wrap it a general type that also stores extra context for tracing, thus keeping error and context separated. 
(see server/parser.go)

The new error behaves just like the original one so it can be used as before.
One caveat is that comparisons have to change from == to errors.Is.
However that can be done gradually when introducing new errors or when modifying an existing one (as I do in this change).
Basically if you find yourself wrapping an error constant with a context, search if there is an instance where this error is compared to a constant and replace == with errors.Is .
(see server/parser_test.go)

If you want to get at the extra information call UnpackIfErrorCtx with any error.
I imagine this to happen in traces. 
To avoid having to modify format arguments I added Error versions to tracing code that simply take an error object.  Again this is optional and is not 100% necessary, but reduces the #calls to UnpackIfErrorCtx. (server/client.go ,  server/log.go)

Example of how this change currently shows up in clients and server tracing.

15:47:40 > ~/test > cat cfg/limit.cfg
listen 127.0.0.1:4222
max_control_line 1
 15:47:47 > ~/test > nats-server -c cfg/limit.cfg &
[1] 6278
[6278] 2020/02/24 15:47:47.686211 [INF] Starting nats-server version 2.1.4
[6278] 2020/02/24 15:47:47.686310 [INF] Git commit [not set]
[6278] 2020/02/24 15:47:47.686519 [INF] Listening for client connections on 127.0.0.1:4222
[6278] 2020/02/24 15:47:47.686526 [INF] Server id is NC5HAQWQCWKFLANIJFWIPBYAFHWS3QX3243OPSQKWXZRM6F5ZIFHSXIE
[6278] 2020/02/24 15:47:47.686528 [INF] Server is ready
 15:47:56 > ~/test > nats-pub -s "127.0.0.1:4222"  wo99999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999977777777777777777777777777777777777777777777777777777777999999999999999999999999999999999999999999999999999999999999999999999999999999ssssssssssssssssssssssssssssssssssssssssssssssssssss999999999999999999999999999999999999999999999999999999999999999999999999999ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss world
[6278] 2020/02/24 15:47:56.999291 [ERR] 127.0.0.1:52910 - cid:1: maximum control line exceeded: State 22, max_control_line 1, Buffer len 508
Server Error: 'maximum control line exceeded'
 15:47:57 > ✘ 1 > ~/test >

Signed-off-by: Matthias Hanel <mh@synadia.com>

### CODE REVIEW

**Comment by matthiashanel** [comment]
This breaks the 1.12 build. adding an alternative to errors.Is is smth I'll address if there is consensus to go ahead with this.

File: server/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2012-2021 The NATS Authors
```
Comment (kozlovic) [inline server/errors_test.go]: Since this is a new file, data should simply be `2020`.

File: server/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
+	if UnpackIfErrorCtx(ErrWrongGateway) != ErrWrongGateway.Error() {
+		t.Fatalf("Error of different type should be processed unchanged")
+	}
+	trace := UnpackIfErrorCtx(e)
+	if !strings.HasPrefix(trace, ErrWrongGateway.Error()) {
+		t.Fatalf("original error needs to remain")
+	}
+	if !strings.HasSuffix(trace, ctx) {
+		t.Fatalf("ctx nneds to be added")
```
Comment (kozlovic) [inline server/errors_test.go]: `needs`

File: server/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
```
Comment (kozlovic) [inline server/errors_test.go]: `needs`

File: server/parser_test.go [hunk 4]
```diff
@@ -15,6 +15,7 @@ package server
 
 import (
 	"bytes"
+	"errors"
```
Comment (kozlovic) [inline server/parser_test.go]: Bump copyright date top of file to "-2020"

File: server/reload.go [hunk 5]
```diff
@@ -643,7 +643,7 @@ func (s *Server) reloadOptions(curOpts, newOpts *Options) error {
 	if err != nil {
 		return err
 	}
-	// Create a context that is used to pass special info that we may need
+	// Create a ctx that is used to pass special info that we may need
```
Comment (kozlovic) [inline server/reload.go]: Looks like a "find and replace" unintended change here

File: server/parser.go [hunk 6]
```diff
@@ -855,9 +855,11 @@ func (c *client) parse(buf []byte) error {
 		// exact at all but the performance hit is too great to be precise, and
 		// catching here should prevent memory exhaustion attacks.
 		if len(c.argBuf) > int(mcl) {
-			c.sendErr(ErrMaxControlLine.Error())
+			err := NewErrorCtx(ErrMaxControlLine, "State %d, max_control_line %d, Buffer len %d",
```
Comment (kozlovic) [inline server/parser.go]: Change copyright date to "-2020"

File: server/log.go [hunk 7]
```diff
@@ -156,6 +156,25 @@ func (s *Server) Errorf(format string, v ...interface{}) {
 	}, format, v...)
 }
 
+// Error logs an error with a scope
```
Comment (kozlovic) [inline server/log.go]: Change copyright date to "-2020"

File: server/errors_test.go (hunk 1 above)
Comment (matthiashanel) [reply server/errors_test.go]: fixed

File: server/errors_test.go (hunk 3 above)
Comment (matthiashanel) [reply server/errors_test.go]: fixed

File: server/reload.go (hunk 5 above)
Comment (matthiashanel) [reply server/reload.go]: yeah, removed

File: server/parser_test.go (hunk 4 above)
Comment (matthiashanel) [reply server/parser_test.go]: fixed

File: server/log.go (hunk 7 above)
Comment (matthiashanel) [reply server/log.go]: fixed

File: server/parser.go (hunk 6 above)
Comment (matthiashanel) [reply server/parser.go]: fixed

File: server/errors_test.go (hunk 2 above)
Comment (matthiashanel) [reply server/errors_test.go]: fixed

**Comment by matthiashanel** [comment]
I added a package gobackcomp that enables errors.Is as ErrorsIs and conditionally compiles based on go version. (go version can't be part of file name like os or arch).
I figured smth. like this might happen more than once, so went with package and exportint functions <orig_package><orig_function>

File: server/gobackcomp/go113_pred.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2020 The NATS Authors
```
Comment (kozlovic) [inline server/gobackcomp/go113_pred.go]: You have 2 files doing the same. Remove one and maybe you can simply name it go112.go, so we know that when we no longer want to support go112.go, we can remove.

Comment (matthiashanel) [reply server/gobackcomp/go113_pred.go]: yes, just realized my commit did not capture renaming it.
current name is go1.13 prev as it applies to previous versions

Comment (kozlovic) [reply server/gobackcomp/go113_pred.go]: As we discussed, I think that we could have the "emulation" code in server/errors.go (and not have this new package) and add a TODO that we can then replace with native code when we no longer support Go 1.12. The issue for supporting older release is that the Go client tests build from source and there in Travis we still have Go 1.11, 1.12 and 1.13.

File: server/errors.go [hunk 9]
```diff
@@ -208,3 +208,78 @@ func (e *processConfigErr) Warnings() []error {
...
+	})
+	if !ok {
+		return nil
+	}
+	return u.Unwrap()
+}
+
+// implements: go 1.13 errors.Is(err, target error) bool
+// TODO replace with native code once we no longer support go1.12
+func IsErr(err, target error) bool {
```
Comment (kozlovic) [inline server/errors.go]: Could we name this ErrorIs() instead ?

Comment (matthiashanel) [reply server/errors.go]: done, Ci failure is unrelated
    events_test.go:978: Expected connection to fail due to max limit

**Pull request merged**